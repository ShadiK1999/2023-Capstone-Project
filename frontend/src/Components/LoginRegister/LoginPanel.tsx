import * as styles from './RegistrationPanel.styles';
import { FC, useCallback, useContext, useState } from 'react';
import { SubmitHandler, useForm } from 'react-hook-form';
import { zodResolver } from '@hookform/resolvers/zod';
import { z } from 'zod';
import { handleLogin } from './Registration.utils';
import { AuthDispatchContext } from '../../Contexts/Auth';
import { useNavigate } from 'react-router-dom';

const validationSchema = z.object({
  email: z.string().nonempty({ message: 'Please enter your login email.' }).email({
    message: 'Must be a valid email.',
  }),
  password: z.string().nonempty({ message: 'Please enter your password.' }),
});

type ValidationSchema = z.infer<typeof validationSchema>;
const resolver = zodResolver(validationSchema);

type ComponentProps = {
  onClose: () => void;
};

const LoginPanel: FC<ComponentProps> = ({ onClose }) => {
  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<ValidationSchema>({ resolver });

  const [submitError, setSubmitError] = useState<string | null>(null);

  const dispatch = useContext(AuthDispatchContext);
  const navigate = useNavigate();

  const onSubmit: SubmitHandler<ValidationSchema> = useCallback(
    async (data) => {
      const error = await handleLogin(data, dispatch, navigate);

      if (error) {
        setSubmitError(error);
      } else {
        onClose();
      }
    },
    [dispatch, navigate, onClose],
  );

  return (
    <div className={styles.Component}>
      <form onSubmit={handleSubmit(onSubmit)}>
        <div className={styles.InputContainer}>
          <input type="text" placeholder="Email" {...register('email')} />
          {errors?.email && <span>{errors.email.message}</span>}
        </div>
        <div className={styles.InputContainer}>
          <input type="password" placeholder="Password" {...register('password')} />
          {errors?.password && <span>{errors.password.message}</span>}
        </div>
        <div className={styles.InputContainer}>
          <label>
            <input type="checkbox" role="checkbox" />
            Remember me
          </label>
        </div>
        <div className={styles.InputContainer}>
          {submitError && <span>{submitError}</span>}
          <button type="submit">Login</button>
        </div>
      </form>
    </div>
  );
};

export { LoginPanel };
