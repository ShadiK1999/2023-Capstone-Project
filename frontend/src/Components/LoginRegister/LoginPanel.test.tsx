import { render, fireEvent } from '@testing-library/react';
import { describe, it, expect } from 'vitest';
import { LoginPanel } from './LoginPanel';
import { MemoryRouter } from 'react-router-dom';

describe('Matches DOM Snapshot', () => {
  it('Matches DOM Snapshot', () => {
    const tree = render(
      <MemoryRouter>
        <LoginPanel onClose={() => void 0} />
      </MemoryRouter>,
    );
    expect(tree).toMatchSnapshot();
  });

  it('Does not allow empty field values', async () => {
    const { getByText, findByText } = render(
      <MemoryRouter>
        <LoginPanel onClose={() => void 0} />
      </MemoryRouter>,
    );
    fireEvent.click(getByText('Login'));
    expect(await findByText('Please enter your login email.')).toBeInTheDocument();
    expect(await findByText('Please enter your password.')).toBeInTheDocument();
  });

  it('Does not allow invalid email format', async () => {
    const { getByText, findByText, getByPlaceholderText } = render(
      <MemoryRouter>
        <LoginPanel onClose={() => void 0} />
      </MemoryRouter>,
    );
    fireEvent.change(getByPlaceholderText('Email'), { target: { value: 'test' } });
    fireEvent.click(getByText('Login'));
    expect(await findByText('Must be a valid email.')).toBeInTheDocument();
  });
});
