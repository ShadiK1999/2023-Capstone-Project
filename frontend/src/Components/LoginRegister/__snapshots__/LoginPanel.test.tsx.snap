// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Matches DOM Snapshot > Matches DOM Snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="f1ur6vce"
      >
        <form>
          <div
            class="fidbxqg"
          >
            <input
              name="email"
              placeholder="Email"
              type="text"
            />
          </div>
          <div
            class="fidbxqg"
          >
            <input
              name="password"
              placeholder="Password"
              type="password"
            />
          </div>
          <div
            class="fidbxqg"
          >
            <label>
              <input
                role="checkbox"
                type="checkbox"
              />
              Remember me
            </label>
          </div>
          <div
            class="fidbxqg"
          >
            <button
              type="submit"
            >
              Login
            </button>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="f1ur6vce"
    >
      <form>
        <div
          class="fidbxqg"
        >
          <input
            name="email"
            placeholder="Email"
            type="text"
          />
        </div>
        <div
          class="fidbxqg"
        >
          <input
            name="password"
            placeholder="Password"
            type="password"
          />
        </div>
        <div
          class="fidbxqg"
        >
          <label>
            <input
              role="checkbox"
              type="checkbox"
            />
            Remember me
          </label>
        </div>
        <div
          class="fidbxqg"
        >
          <button
            type="submit"
          >
            Login
          </button>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
